src/types.ts

```diff
@@ -1,4 +1,26 @@
 import {ReporterType} from './reporter'
+import {PackageSpec, ResolveOptions, ResolveResult} from './resolve';
+
+export type LifecycleHooks = {
+
+  /**
+   * Executes before the pnpm's default resolution takes action.
+   *
+   * If this method returns a resolution then it is used, otherwise pnpm's
+   * default resolution algorithm will be used.
+   */
+  packageWillResolve?: (spec: PackageSpec, opts: ResolveOptions) => Promise<ResolveResult | null>,
+
+  // TODO: add more lifecycle hooks
+  //
+  // packageDidResolve
+  //
+  // packageWillFetch
+  // packageDidFetch
```

Comment:
`packageDidFetch(spec: PackageSpec, target: string)` probably can be used to check if `package.json` is exists and generate one if needed.